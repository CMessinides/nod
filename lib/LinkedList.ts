export default function* LinkedList<
	L extends Record<string | symbol | number, any>,
	I extends keyof L,
	P extends keyof L
>(list: L[], idKey: I, prevIdKey: P) {
	if (list.length === 0) return;

	let currentId = null;

	// Selectors to make ID access more readable
	const getId = (item: L) => item[idKey];
	const getPrevId = (item: L) => item[prevIdKey];

	// Create a hash table mapping
	// each item's previous ID to its index in
	// the unsorted list.
	const map = list.reduce(
		(map, item, index) => {
			map[getPrevId(item)] = index;
			return map;
		},
		{} as Record<string | symbol | number, number>
	);

	for (let i = 0; i < list.length; i++) {
		const nextItem = list[map[currentId]];
		yield nextItem;
		currentId = getId(nextItem);
	}
}
