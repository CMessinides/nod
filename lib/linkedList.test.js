import LinkedList from "./LinkedList";

it("should order a linked list", () => {
	const unsortedList = [
		{
			id: 5,
			prevId: 7
		},
		{
			id: 7,
			prevId: null
		},
		{
			id: 3,
			prevId: 4
		},
		{
			id: 6,
			prevId: 5
		},
		{
			id: 4,
			prevId: 6
		}
	];

	expect(
		Array.from(LinkedList(unsortedList, "id", "prevId"))
	).toMatchSnapshot();
});
