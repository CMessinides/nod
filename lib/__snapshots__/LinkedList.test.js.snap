// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should order a linked list 1`] = `
Array [
  Object {
    "id": 7,
    "prevId": null,
  },
  Object {
    "id": 5,
    "prevId": 7,
  },
  Object {
    "id": 6,
    "prevId": 5,
  },
  Object {
    "id": 4,
    "prevId": 6,
  },
  Object {
    "id": 3,
    "prevId": 4,
  },
]
`;
